### PR TITLE
refactor(app): Refactor labware calibration flow to use new modal(s)

### PR DESCRIPTION
**Pull request by mcous**
## overview

This PR is pulling double duty as a ticket. It implements the following screens:

screenshot 2018-02-07 19 09 39

screenshot 2018-02-07 19 09 44

screenshot 2018-02-07 19 09 49

screenshot 2018-02-07 19 10 05

In combination with #677 and some additional polish, this will close out the labware calibration milestone.

## changelog

- Refactor: removed old labware calibration prompts in favor of new modals from screens above
- Refactor: tiprack calibration has a new procedure:
    1. Move to top of tip
    2. Ask user to jog if needed
    3. Pick up tips and ask user if pickup was successeful
        * If yes: done
        * If no: drop tip, move to top of tip, goto (2)
- Fix: fixed trash no longer shows up in labware calibration list

### calibration flow in redux action creators

**Previously**

Tipracks:

1. `pickupAndHome`
2. Maybe pickup fails
    1. `dropTipAndHome`
    2. `jog`
    3. `updateOffset` - Would update container offset _and_ then pickup and home
    4. GOTO (1)
3. `confirmTiprack` - Drops tip unless tiprack is last

Non-tipracks:

1. `moveTo`
2. If movement is correct, `confirmLabware`
3. Else, `jog`, then `updateOffset`

**New hotness**

Tipracks:

1. `moveTo`
2. Maybe `jog`
3. `pickupAndHome` - Updates container offset, then picks up and homes
4. Maybe pickup fails
    1. `dropTipAndHome` - Drops tip, homes, and then calls `move_to`
    2. GOTO (2)
5. `confirmTiprack`

Non-topracks:

1. `moveTo`
2. Maybe `jog`
3. `updateOffset`
    * TODO: rename to `confirmLabware` and remove old `confirmLabware`

## review requests

This is a big one, sorry. Unfortunately this flow only really makes sense with all the pieces in place. I split out as much as I could in earlier PRs.

Because of the behavioral changes, robot testing will be important here. @btmorr, just as a reminder we had some API stuff that testing this PR brought up:

* Updated labware positions are not reflected in `calibration_manager.move_to` calls
* Unused pipettes do not retract during tip pickups

### CODE REVIEW

File: app/src/components/LabwareCalibration/ConfirmPositionContents.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+  const {slot, isTiprack, calibrator: {mount}} = ownProps
+
+  const jogButtons = JOG_BUTTONS.map((button) => {
+    const {name, axis, direction, arrow} = button
+    const onClick = () => {
+      dispatch(robotActions.jog(mount, axis, direction))
+    }
+
+    return {name, arrow, onClick}
+  }, {})
```
Comment (mcous) [inline app/src/components/LabwareCalibration/ConfirmPositionContents.js]: Should just be `})` (this is an accidental remnant of when this call was a reduce in an earlier draft.

File: app/src/components/LabwareCalibration/ConfirmPositionContents.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
+// @flow
+// JogControls modal contents for ConfirmationModal
```
Comment (mcous) [inline app/src/components/LabwareCalibration/ConfirmPositionContents.js]: Out of date comment

File: app/src/components/LabwareCalibration/ConfirmPositionContents.js [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  )
+}
+
+const JOG_BUTTONS: Array<{
+  name: JogButtonName,
+  axis: Axis,
+  direction: Direction,
+  arrow: IconName
+}> = [
+  {name: 'left', axis: 'x', direction: -1, arrow: CHEVRON_LEFT},
```
Comment (mcous) [inline app/src/components/LabwareCalibration/ConfirmPositionContents.js]: Does this icon mapping really belong here in the container? It should probably go in the JogControls component (axis and direction data still belong here as they're logic)

File: app/src/components/LabwareCalibration/JogControls.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
+// @flow
+// jog controls container for ConfirmPositionContents
```
Comment (mcous) [inline app/src/components/LabwareCalibration/JogControls.js]: Out of date comment (not a container)

File: app/src/components/deck/ConnectedLabwareItem.js [hunk 5]
```diff
@@ -33,46 +30,39 @@ function mapStateToProps (state, ownProps) {
   } = labware
 
   const current = slotName === selectedSlot
-  const highlighted = (
-    (current && !confirmed) ||
-    (selectedLabware.confirmed && nextLabware && slotName === nextLabware.slot)
-  )
+  const highlighted = current
```
Comment (mcous) [inline app/src/components/deck/ConnectedLabwareItem.js]: It looks like `current` isn't used anywhere else and can be removed

File: app/src/components/deck/LabwareItem.js [hunk 6]
```diff
@@ -12,11 +14,13 @@ import {
...
 } from '@opentrons/components'
 
+import styles from './deck.css'
+
 type LabwareItemProps = {
   highlighted?: boolean,
   confirmed?: boolean,
   isMoving?: boolean,
-  canRevisit?: boolean, // if true, wrap labware in a Link
+  disabled?: boolean,
```
Comment (mcous) [inline app/src/components/deck/LabwareItem.js]: A bunch of these parameters don't seem optional

File: app/src/components/deck/deck.css [hunk 7]
```diff
@@ -29,3 +29,12 @@ button.btn_calibrate {
   background-color: var(--c-blue);
   color: white;
 }
+
+.disabled {
+  opacity: 0.4;
```
Comment (mcous) [inline app/src/components/deck/deck.css]: I should've been less lazy and done the proper color thing

File: app/src/pages/SetupDeck.js [hunk 8]
```diff
@@ -21,27 +19,15 @@ type Props = ContextRouter & StateProps
 export default withRouter(connect(mapStateToProps)(SetupDeckPage))
 
 function SetupDeckPage (props: Props) {
-  const {location: {pathname}, match: {url, path, params: {slot}}} = props
-
-  // redirect to review modal if deck not ready and we're not already there
-  const redirectToReview = (
```
Comment (mcous) [inline app/src/pages/SetupDeck.js]: This whole `<Redirect>` thing turned out to be cleverness for cleverness' sake. My bad y'all; I got too excited about a toy I hadn't really used yet

File: app/src/pages/SetupDeck.js [hunk 9]
```diff
@@ -21,27 +19,15 @@ type Props = ContextRouter & StateProps
...
       <SessionHeader subtitle='Setup Deck' />
-      <LabwareCalibrationInfo slot={slot} />
-      <CalibrateDeck slot={slot} />
-      <Route path={`${path}/jog`} render={() => (
-        <ConnectedJogModal slot={slot} />
-      )} />
-      <Route path={`${path}/review`} render={() => (
-        <ReviewDeckModal slot={slot} closeUrl={url} />
-      )} />
+      <LabwareCalibration slot={slot} url={url} />
```
Comment (mcous) [inline app/src/pages/SetupDeck.js]: I put the (route driven) calibration modals all entirely within the LabwareCalibration component. @Kadee80 and @IanLondon how do you feel about this? I could yank them out to here. I kind of like where they are in that then _all_ the labware calibration logic is in one directory

File: app/src/robot/api-client/client.js [hunk 10]
```diff
@@ -144,12 +144,18 @@ export default function client (dispatch) {
...
 
+  // saves container offset and then attempts a tip pickup
   function pickupAndHome (state, action) {
     const {payload: {mount, slot}} = action
     const instrument = {_id: selectors.getInstrumentsByMount(state)[mount]._id}
     const labware = {_id: selectors.getLabwareBySlot(state)[slot]._id}
 
-    remote.calibration_manager.pick_up_tip(instrument, labware)
+    // FIXME(mc, 2017-10-05): DEBUG CODE
+    // return setTimeout(() => dispatch(actions.pickupAndHomeResponse()), 1000)
```
Comment (mcous) [inline app/src/robot/api-client/client.js]: Note to self: make an actual mock server that can be enabled by an environment variable

File: app/src/robot/api-client/client.js [hunk 11]
```diff
@@ -172,11 +182,22 @@ export default function client (dispatch) {
     const labware = {_id: selectors.getLabwareBySlot(state)[slot]._id}
 
     if (selectors.getUnconfirmedTipracks(state).length === 1) {
-      return dispatch(actions.confirmTiprackResponse(null, true))
+      dispatch(actions.confirmTiprackResponse(null, true))
+      dispatch(push(`/setup-deck/${slot}`))
```
Comment (mcous) [inline app/src/robot/api-client/client.js]: Yeah, we've had more routes leak into the API client. This seems like a job for middleware...

Some sort of standard thunk / promise middleware might help, especially with the continuing addition of HTTP endpoints to the API (hello `window.fetch`)

File: app/src/robot/test/actions.test.js [hunk 12]
```diff
@@ -337,15 +337,15 @@ describe('robot actions', () => {
...
+      payload: {}
     }
     const failure = {
       type: 'robot:UPDATE_OFFSET_FAILURE',
       error: true,
       payload: new Error('AH')
     }
 
-    expect(actions.updateOffsetResponse(null, true)).toEqual(success)
+    expect(actions.updateOffsetResponse(null)).toEqual(success)
```
Comment (mcous) [inline app/src/robot/test/actions.test.js]: Unnecessary `null`

File: app/src/components/LabwareCalibration/styles.css [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+  text-align: center;
+}
+
+.jog_label_xy {
+  grid-column: 2 / 5;
+  grid-row: 3;
+}
+
+.jog_label_z {
+  grid-column: 7;
```
Comment (mcous) [inline app/src/components/LabwareCalibration/styles.css]: I don't think I understand row and column these callouts nor `justify-self` and `align-self`. @Kadee80 could you help me out here / check my work?

File: app/src/components/LabwareCalibration/ConfirmModalContents.js [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+type OwnProps = Labware
+
+type StateProps = {
+  calibrator: Instrument
+}
+
+type Props = OwnProps & StateProps
+
+const CONTENTS_BY_STATUS: {
```
Comment (mcous) [inline app/src/components/LabwareCalibration/ConfirmModalContents.js]: This object probably makes a lot more sense as an if-else ladder or switchcase

File: app/src/robot/api-client/client.js [hunk 15]
```diff
@@ -198,10 +211,16 @@ export default function client (dispatch) {
...
+    // return setTimeout(() => {
+    //   dispatch(actions.moveToResponse())
+    //   dispatch(push(`/setup-deck/${slot}/confirm`))
+    // }, 1000)
 
     remote.calibration_manager.move_to(instrument, labware)
-      .then(() => dispatch(actions.moveToResponse()))
+      .then(() => {
+        dispatch(actions.moveToResponse())
+        dispatch(push(`/setup-deck/${slot}/confirm`))
```
Comment (mcous) [inline app/src/robot/api-client/client.js]: I'm a little torn. @Kadee80 do we prefer:

a) This push over here in the webworker client?

or

b) A weird `<Redirect>` over in `<LabwareConfirmation>`?

File: app/src/components/LabwareCalibration/styles.css (hunk 13 above)
Comment (Kadee80) [reply app/src/components/LabwareCalibration/styles.css]: I could have been a bit redundant here, but I believe the the align and justify is for centering the text horizontally and vertically (might also be a ghost of CSS past).

The row col stuff is shorthand. 
`grid-template-rows` and `grid-template-columns` generated 2 rows and 6 columns of 3rem x 3rem grid template areas. the auto value is for filling the extra space and centering. 

If an item is only 1 grid area wide or tall you use one number.
`grid-column: 2/5` shorthand syntax declares the start and end values at once, so the label starts at beginning 2 and end at the beginning of 5.

I'll check around this file locally, but it renders fine.

File: app/src/robot/api-client/client.js (hunk 15 above)
Comment (Kadee80) [reply app/src/robot/api-client/client.js]: I think i saw another comment about you not being so keen on redirect?
Webworker client seems like it should stick to API stuff but this seems like a possibly worthy 1 off. 
Would the alternative be moveToResponse updates something in state, and that triggers a redirect?

I'm a little torn as well. I could argue for either. I guess it comes down to how comfortable we are with hiding router actions in client.

File: app/src/pages/SetupDeck.js (hunk 9 above)
Comment (Kadee80) [reply app/src/pages/SetupDeck.js]: I'm fine with it for now. If it becomes an issue we can just do some charlie work.

File: app/src/robot/api-client/client.js (hunk 15 above)
Comment (mcous) [reply app/src/robot/api-client/client.js]: Yeah, the moveToResponse updates the labware status from `moving-to-slot` to `over-slot`, so we could use that to trigger the redirect. But we'd also need to check if we're already redirected, and that's where that logic starts to annoy me.

Seems like we're on the same page. Gonna keep this one in place until we can figure out a more general solution (seeing as we've got other `push`es in the client anyway.

**Pull request merged**